### ISSUE TITLE
Improve documentation of DynamicTest lifecycle

### ISSUE THREAD

**Issue by littleclay** [junit-team/junit5#386]
This might not be a bug and may be as intended, but kinda hit me by surprise as I don't see anything about it in the user guide or java docs. It appears that when you generate `DynamicTest`s they are different from regular `@Test` methods in _not_ generating separate instances of the test class containing them for each method. For example, the following code:

``` java
public class FooTest {

    @Test
    public void foo() {
        System.out.println( this );
    }

    @TestFactory
    public Stream< DynamicTest > testFactory() {
        return Stream.of( "A", "B" )
                     .map( str -> DynamicTest.dynamicTest( "testFromFactory" + str + "()",
                                                           () -> System.out.println( str + " - " + this ) ) );
    }
}
```

might have sample output such as this:

```
Running org.littleclay.FooTest
org.littleclay.FooTest@5cc7c2a6
A - org.littleclay.FooTest@32d992b2
B - org.littleclay.FooTest@32d992b2
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.032 sec - in org.littleclay.FooTest
```

Note that the method `foo` has a separate instance of `FooTest` during it's execution but that both `A` and `B` tests share the same `FooTest` instance. It seems to me like this is a bit of a "gotcha" in using dynamic tests and think it should probably be spelled out more clearly if this is a necessary by-product of this type of test.

In a similar vein, it seems that the extension callbacks with dynamic tests have the same "method" which threw me a bit by surprise when following the examples regarding storing and retrieving state with extensions. Again, it makes sense with how I'm guessing these are generated, but if they can't be made more isolated, I definitely think a bit more information regarding life-cycles and possible contention points would be nice.

It is probably _less_ of an issue if these types of dynamic tests can never be run in parallel, but I already see a few issues with my initial use case for them for creating parameterized tests if they cannot be (1) run in parallel and (2) be more isolated.
## Related Issues
- #14 
- #378
- #393

**Comment by sbrannen** [comment]
These are in fact not bugs but rather direct consequences of how lambda expressions are implemented in Java.

> It appears that when you generate `DynamicTest`s they are different from regular `@Test` methods in *not* generating separate instances of the test class containing them for each method.

That's correct. This is how Java works and actually has nothing to do with JUnit per se. As such, there is nothing that JUnit can do to influence this. The User Guide assumes developers understand the semantics of lambda expressions in Java, but admittedly lambdas are new to a lot of people. So it would probably help to at least point this out in the User Guide.

> In a similar vein, it seems that the extension callbacks with dynamic tests have the same "method" which threw me a bit by surprise when following the examples regarding storing and retrieving state with extensions.

This is actually addressed in the [User Guide](http://junit.org/junit5/docs/current/user-guide/#writing-tests-dynamic-tests) in the _Dynamic Test Lifecycle_ warning:

> The execution lifecycle of a dynamic test is quite different than it is for a standard `@Test` case. Specifically, there are not any lifecycle callbacks for dynamic tests. This means that `@BeforeEach` and `@AfterEach` methods and their corresponding extension callbacks are not executed for dynamic tests. In other words, if you access fields from the test instance within a lambda expression for a dynamic test, those fields will not be reset by callback methods or extensions between the execution of dynamic tests generated by the same `@TestFactory` method. 

The key part is: **there are not any lifecycle callbacks for dynamic tests**.

There are in fact lifecycle callbacks for the `@TestFactory` method, but that is not the same thing.

> I definitely think a bit more information regarding life-cycles and possible contention points would be nice.

I agree and will therefore update the title of this issue accordingly.

> It is probably less of an issue if these types of dynamic tests can never be run in parallel, but I already see a few issues with my initial use case for them for creating parameterized tests if they cannot be (1) run in parallel and (2) be more isolated.

Some of these issues will be addressed in #378; others will be addressed in #14.

**Comment by littleclay** [comment]
Doh, I do remember reading that about callbacks and dynamic tests now. Thank you for responding!

**Comment by littleclay** [comment]
Also, that totally makes sense about lambdas and test class instances now that I think about it. That'll teach me to try to plan out testing models at night when I'm tired :-)

**Comment by sbrannen** [comment]
No worries: dynamic tests with lambdas equate to new territory for all of us. So there's room for improvement, both in terms of documentation and features.

**Comment by marcphilipp** [comment]
Closed as duplicate of #374.
